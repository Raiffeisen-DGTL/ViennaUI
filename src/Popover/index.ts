export { Popover, type PopoverProps } from './Popover';
