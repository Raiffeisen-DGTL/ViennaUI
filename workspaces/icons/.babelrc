{
  "presets": ["@babel/env", "@babel/typescript", "@babel/react"],
  "env": {
    "esm": {
      "presets": [["@babel/env", { "modules": false }]]
    }
  }
}
