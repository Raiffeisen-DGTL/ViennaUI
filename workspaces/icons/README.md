# Библиотека иконок

###### Raiffeisenbank Design System

Коллекция иконок для использования в интерфейсах. Иконки представлены в виде React компонентов и оригинальных векторных (`*.svg`) изображений. Изображения напрямую можно использовать из пакета.
