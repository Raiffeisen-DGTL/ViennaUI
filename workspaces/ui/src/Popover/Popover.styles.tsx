import styled, { css } from 'styled-components';
import { getPresets } from 'vienna.ui-primitives';

const message = getPresets('popover.message', {
    base: null,
    design: 'design',
    custom: null,
});

const header = getPresets('popover.header', {
    base: null,
    custom: null,
});

const content = getPresets('popover.content', {
    base: null,
    custom: null,
});

const closeIcon = getPresets('popover.close', {
    base: null,
    custom: null,
});

const defenition = getPresets('popover.defenition', {
    base: null,
    custom: null,
});

const wrapper = getPresets('popover.wrapper', {
    custom: null,
});

export const Wrapper = styled.span`
    outline: none;
    position: relative;
    display: inline-flex;
    ${wrapper.custom}
`;

export const Message = styled.div<{
    active?: boolean;
    toggle?: boolean;
    design?: 'dark' | 'light';
    left?: any;
    top?: any;
    right?: any;
    width?: number;
    bottom?: any;
}>`
    position: absolute;
    display: flex;
    flex-direction: column;
    opacity: 0;
    ${({ width }) => `width: ${width}px;`}
    ${({ width }) => `min-width: ${width}px;`}
    ${message.base}
    ${message.design}
    ${message.custom}

    ${({ top, left, right, bottom }) =>
        css`
            top: ${top}px;
            left: ${left}px;
            right: ${right}px;
            bottom: ${bottom}px;
        `}

    ${({ active }) =>
        !active &&
        css`
            display: none;
        `}


    ${({ toggle }) =>
        toggle &&
        css`
            opacity: 1;
        `}
`;

export const Header = styled.div`
    ${header.base}
    ${header.custom}
`;

export const Content = styled.div`
    ${content.base}
    ${content.custom}
`;

export const CloseIcon = styled.div`
    position: absolute;
    ${closeIcon.base}
    ${closeIcon.custom}
`;

export const Defenition = styled.span`
    ${defenition.base}
`;

export const ScrollLocker = styled.div`
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background: transparent;
    z-index: 500;
`;
