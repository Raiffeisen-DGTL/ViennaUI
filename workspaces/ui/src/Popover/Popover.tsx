import React, { useState, useRef, useCallback, useEffect, useMemo, useLayoutEffect } from 'react';
import { useOutsideClick, getScrollParent } from 'vienna.react-use';
import { Close } from 'vienna.icons';
import { Wrapper, Message, Header, Content, CloseIcon, ScrollLocker } from './Popover.styles';

export interface PopoverProps {
    [key: string]: any;
    id?: string;
    design?: 'dark' | 'light';
    anchor?: 'left' | 'right' | 'top' | 'bottom' | 'center' | 'auto';
    header?: React.ReactNode | string;
    content?: React.ReactNode | string;
    /** Блокировать прокрутку экрана если элемент активен */
    noScroll?: boolean;
    /** Включить поддержку событий клавиатуры (для Popover = false, для Hint = true) */
    allowKeyboardEvents?: boolean;
    /** Ширина элемента (250px по умолчанию) */
    width?: number;
    /** Отключает иконку закрытия */
    noClose?: boolean;
    /* Обработчик открытия элемента */
    onOpen?: () => void;
    /* Обработчик закрытия элемента */
    onClose?: () => void;
}

interface Pos {
    left?: number;
    right?: number;
    top?: number;
    bottom?: number;
}

interface Ref {
    open: () => void;
    close: () => void;
    isOpen: boolean;
}

const SHIFT = 8;

export const Popover: React.FC<PopoverProps> = React.forwardRef(
    (props: React.PropsWithChildren<PopoverProps>, ref: React.Ref<Ref>) => {
        const {
            id,
            children,
            content,
            header,
            anchor = 'auto',
            design = 'light',
            width = 250,
            allowKeyboardEvents = false,
            noScroll,
            noClose,
            onOpen,
            onClose,
            ...attrs
        } = props;

        const [active, setActive] = useState(false);
        const [toggle, setToggle] = useState(false);
        const [pos, setPos] = useState<Pos>({ top: 22, left: 0 });

        const messageRef = useRef<HTMLDivElement>(null);
        const wrapperRef = useRef<HTMLDivElement>(null);

        const scrollableParent: any = useMemo(
            () => ({
                current: null,
            }),
            []
        );

        const close = useCallback(() => {
            setToggle(false);
            setTimeout(() => setActive(false), 200);
            const parent = scrollableParent.current;
            if (parent && noScroll) {
                parent.style.overflow = '';
            }
            onClose?.();
        }, [scrollableParent, noScroll, onClose]);

        const open = useCallback(() => {
            setActive(true);
            const parent = scrollableParent.current;
            if (parent && noScroll) {
                const temp = parent.scrollTop;
                parent.style.overflow = 'hidden';
                parent.scrollTop = temp;
            }
            onOpen?.();
        }, [scrollableParent, noScroll, onOpen]);

        useOutsideClick(messageRef, close, { isEnabled: active });

        const handleFocus = useCallback(() => {
            if (allowKeyboardEvents) {
                open();
            }
        }, [allowKeyboardEvents, open]);

        const handleKeyDown = useCallback(
            (e) => {
                if (allowKeyboardEvents) {
                    e.stopPropagation();
                    if (e.key === 'Enter' && !active) {
                        open();
                    }
                    if (e.key === 'Escape' && active) {
                        close();
                    }
                }
            },
            [active, close, allowKeyboardEvents, open]
        );

        useEffect(() => {
            if (ref) {
                const refObj = {
                    open,
                    close,
                    get isOpen() {
                        return toggle;
                    },
                };
                if (typeof ref === 'function') {
                    ref(refObj);
                } else {
                    (ref as any).current = refObj;
                }
            }

            return () => {
                if (ref) {
                    if (typeof ref === 'function') {
                        ref(null);
                    } else {
                        (ref as any).current = null;
                    }
                }
            };
        }, [ref, active, close, open]);

        useEffect(() => {
            const cur = messageRef.current;
            scrollableParent.current = getScrollParent(cur);
            return () => {
                const parent = scrollableParent.current;
                if (parent && noScroll) {
                    parent.style.overflow = '';
                }
            };
        }, [scrollableParent, noScroll]);

        useLayoutEffect(() => {
            const cur = messageRef.current;
            const component = wrapperRef.current;
            if (cur && component && active) {
                cur.style.visibility = 'hidden';
                const rect = cur.getBoundingClientRect();
                const comRect = component.getBoundingClientRect();
                const parent = getScrollParent(cur);
                const parentRect = parent.getBoundingClientRect();

                switch (anchor) {
                    case 'left': {
                        setPos({
                            right: comRect.width + SHIFT,
                            [rect.top + rect.height < parentRect.top ? 'top' : 'bottom']: 0,
                        });
                        break;
                    }
                    case 'top': {
                        setPos({
                            bottom: comRect.height + SHIFT,
                            [rect.left - rect.width > parentRect.left ? 'right' : 'left']: 0,
                        });
                        break;
                    }
                    case 'right': {
                        setPos({
                            left: comRect.width + SHIFT,
                            [rect.top + rect.height < parentRect.bottom ? 'top' : 'bottom']: 0,
                        });
                        break;
                    }
                    case 'bottom': {
                        setPos({
                            top: comRect.height + SHIFT,
                            [rect.left - rect.width > parentRect.left ? 'right' : 'left']: 0,
                        });
                        break;
                    }
                    case 'center': {
                        setPos({
                            left: comRect.width / 2 - rect.width / 2,
                            [rect.top + rect.height < parentRect.bottom ? 'top' : 'bottom']: comRect.height + SHIFT,
                        });
                        break;
                    }
                    case 'auto':
                        {
                            // Определяем возможные валидные смещенния 0 - смещение нвозможно, N + SHIFT - смещение доступно
                            const result = {
                                left:
                                    comRect.left + comRect.width + rect.width + SHIFT < parentRect.right
                                        ? comRect.width + SHIFT
                                        : 0,
                                right:
                                    comRect.right - comRect.width - rect.width - SHIFT > parentRect.left
                                        ? comRect.width + SHIFT
                                        : 0,
                                top:
                                    comRect.top + comRect.height + rect.height + SHIFT < parentRect.bottom
                                        ? comRect.height + SHIFT
                                        : 0,
                                bottom:
                                    comRect.bottom - comRect.height - rect.height - SHIFT > parentRect.top
                                        ? comRect.height + SHIFT
                                        : 0,
                            };

                            // Ищем доступное смещение исходя из того что Popover позиционируется всегода со смещением 2comRect.width или comRect.height в одном из четырех направлений
                            // Первым проверяем выход вниз и на лево и так далее по цепочке выход возможен если сумма двух составляющих смещений равна 4comRect.width + comRect.height + [Два сдвига]
                            // Т.е. мы можем сместить Popover по этим направлениям
                            // Затем обнуляем нужное смещение и удаляем лишнии, которые не участвуют в позиционировании
                            const sum = comRect.width + comRect.height + SHIFT * 2;

                            if (result.left + result.top === sum) {
                                result.left = 0;
                                delete result.right;
                                delete result.bottom;
                                setPos(result);
                                break;
                            }
                            if (result.left + result.bottom === sum) {
                                result.left = 0;
                                delete result.right;
                                delete result.top;
                                setPos(result);
                                break;
                            }
                            if (result.right + result.top === sum) {
                                result.right = 0;
                                delete result.left;
                                delete result.bottom;
                                setPos(result);
                                break;
                            }
                            if (result.right + result.bottom === sum) {
                                result.right = 0;
                                delete result.left;
                                delete result.top;
                                setPos(result);
                                break;
                            }
                        }

                        setPos({});
                        break;
                }

                setToggle(true);
                cur.style.visibility = '';
            }
        }, [active, anchor, scrollableParent]);

        const scrollLockerClickHandle = useCallback((e) => {
            e.stopPropagation();
        }, []);

        return (
            <Wrapper
                ref={wrapperRef}
                id={id}
                tabIndex={0}
                onFocus={handleFocus}
                onKeyDown={handleKeyDown}
                onClick={active ? undefined : open}
                {...attrs}>
                {children}
                <Message ref={messageRef} active={active} toggle={toggle} design={design} width={width} {...pos}>
                    {!noClose && (
                        <CloseIcon>
                            <Close size='s' onClick={close} />
                        </CloseIcon>
                    )}
                    {header && <Header>{header}</Header>}
                    <Content>{content}</Content>
                </Message>
                {active && noScroll && <ScrollLocker onClick={scrollLockerClickHandle} />}
            </Wrapper>
        );
    }
);

Popover.displayName = 'Popover';
export default Popover;
