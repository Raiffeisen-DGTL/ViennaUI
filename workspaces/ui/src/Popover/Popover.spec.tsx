import React from 'react';
import { Close } from 'vienna.icons';
import { Popover, PopoverProps } from './Popover';
import { Content, Message } from './Popover.styles';
import { Button } from '..';

test('Popover', () => {
    const snap = snapshot.render(
        <Popover
            design='dark'
            anchor='bottom'
            header='This is header'
            content='Ut ut dui non ipsum pharetra aliquam id in libero. Aenean placerat leo nec ex pharetra finibus. Nunc id turpis eu risus rhoncus cursus quis efficitur sapien. Donec vitae arcu at lectus placerat fringilla.'>
            <Button>Click me</Button>
        </Popover>
    );

    expect(snap).toMatchSnapshot();
});

test('Popover w/ design', () => {
    const designs: PopoverProps['design'][] = ['dark', 'light'];

    designs.forEach((design) => {
        const snap = snapshot.render(
            <Popover
                design={design}
                anchor='bottom'
                header='This is header'
                content='Ut ut dui non ipsum pharetra aliquam id in libero. Aenean placerat leo nec ex pharetra finibus. Nunc id turpis eu risus rhoncus cursus quis efficitur sapien. Donec vitae arcu at lectus placerat fringilla.'>
                <Button>Click me</Button>
            </Popover>
        );

        expect(snap).toMatchSnapshot();
    });
});

test('Popover /w focus', () => {
    const snap = snapshot.shallow(
        <Popover
            design='dark'
            anchor='bottom'
            header='This is header'
            content='Ut ut dui non ipsum pharetra aliquam id in libero. Aenean placerat leo nec ex pharetra finibus. Nunc id turpis eu risus rhoncus cursus quis efficitur sapien. Donec vitae arcu at lectus placerat fringilla.'>
            <Button>Click me</Button>
        </Popover>
    );

    const button = snap.find(Button);
    button.simulate('focus');
    const content = snap.find(Content);

    expect(content.exists()).toEqual(true);
});

test('Popover /w close', () => {
    const snap = snapshot.shallow(
        <Popover
            design='dark'
            anchor='bottom'
            header='This is header'
            content='Ut ut dui non ipsum pharetra aliquam id in libero. Aenean placerat leo nec ex pharetra finibus. Nunc id turpis eu risus rhoncus cursus quis efficitur sapien. Donec vitae arcu at lectus placerat fringilla.'>
            <Button>Click me</Button>
        </Popover>
    );

    const button = snap.find(Button);
    button.simulate('focus');
    const close = snap.find(Close);
    close.simulate('click');

    const message = snap.find(Message);
    expect(message.props().toggle).toEqual(false);
});

test('Popover w/ any props', () => {
    const props = {
        id: 'id',
        title: 'title',
    };

    const component = <Popover {...props} />;

    const snap = snapshot.render(component);
    expect(snap).toMatchSnapshot();

    const wrapper = snapshot.mount(component);
    for (const [key, value] of Object.entries(props)) {
        expect(wrapper.props()[key]).toEqual(value);
    }
});
