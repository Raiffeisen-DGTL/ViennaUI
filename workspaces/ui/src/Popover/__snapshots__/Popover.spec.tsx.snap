// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover 1`] = `
.c0 {
  outline: none;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  opacity: 0;
  width: 250px;
  min-width: 250px;
  box-shadow: 0 8px 32px 0 rgba(0,0,0,0.08);
  border-radius: 8px;
  padding-top: 12px;
  padding-right: 28px;
  padding-bottom: 12px;
  padding-left: 12px;
  z-index: 1000;
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
  background-color: #2B2D33;
  color: #FFFFFF;
  top: 22px;
  left: 0px;
  right: px;
  bottom: px;
  display: none;
}

.c5 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
  font-weight: 700;
  margin-bottom: 4px;
}

.c6 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
}

.c4 {
  position: absolute;
  top: 12px;
  right: 12px;
  cursor: pointer;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  outline: 0;
  position: relative;
  border-radius: 8px;
  border-color: transparent;
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.3s ease,color 0.3s ease,border-color 0.3s ease;
  transition: background-color 0.3s ease,color 0.3s ease,border-color 0.3s ease;
  font-size: 14px;
  line-height: 22px;
  padding: 0 20px;
  height: 36px;
  min-width: 36px;
  color: #FFFFFF;
  background-color: #2B2D33;
}

.c1:active {
  top: 1px;
}

.c1:hover {
  background-color: #000000;
  color: #FFFFFF;
}

.c2 {
  display: inline-block;
  vertical-align: top;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2:not(:last-of-type) {
  margin-right: 8px;
}

.c2 > svg {
  display: block;
}

<span
  class="c0"
  data-marker="Popover"
  tabindex="0"
>
  <button
    class="c1"
    data-marker="Button"
    type="button"
  >
    <span
      class="c2"
      data-marker="ButtonComponent.content.ContentWrapper"
    >
      Click me
    </span>
  </button>
  <div
    class="c3"
    data-marker="Popover.Message"
    width="250"
  >
    <div
      class="c4"
      data-marker="Popover.CloseIcon"
    >
      <svg
        data-marker="Popover.Close"
        fill="currentColor"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs
          data-marker="Close.defs"
        />
        <path
          d="M10.586 12L4 18.586V20h1.414L12 13.414 18.586 20H20v-1.414L13.414 12 20 5.414V4h-1.414L12 10.586 5.414 4H4v1.414z"
          data-marker="Close.path"
        />
      </svg>
    </div>
    <div
      class="c5"
      data-marker="Popover.Header"
    >
      This is header
    </div>
    <div
      class="c6"
      data-marker="Popover.Content"
    >
      Ut ut dui non ipsum pharetra aliquam id in libero. Aenean placerat leo nec ex pharetra finibus. Nunc id turpis eu risus rhoncus cursus quis efficitur sapien. Donec vitae arcu at lectus placerat fringilla.
    </div>
  </div>
</span>
`;

exports[`Popover w/ any props 1`] = `
.c0 {
  outline: none;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  opacity: 0;
  width: 250px;
  min-width: 250px;
  box-shadow: 0 8px 32px 0 rgba(0,0,0,0.08);
  border-radius: 8px;
  padding-top: 12px;
  padding-right: 28px;
  padding-bottom: 12px;
  padding-left: 12px;
  z-index: 1000;
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
  background-color: #FFFFFF;
  color: #2B2D33;
  top: 22px;
  left: 0px;
  right: px;
  bottom: px;
  display: none;
}

.c3 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
  font-weight: 700;
  margin-bottom: 4px;
}

.c4 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
}

.c2 {
  position: absolute;
  top: 12px;
  right: 12px;
  cursor: pointer;
}

<span
  class="c0"
  data-marker="Popover_3"
  id="id"
  tabindex="0"
  title="title"
>
  <div
    class="c1"
    data-marker="Popover.Message"
    width="250"
  >
    <div
      class="c2"
      data-marker="Popover.CloseIcon"
    >
      <svg
        data-marker="Popover.Close"
        fill="currentColor"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs
          data-marker="Close.defs"
        />
        <path
          d="M10.586 12L4 18.586V20h1.414L12 13.414 18.586 20H20v-1.414L13.414 12 20 5.414V4h-1.414L12 10.586 5.414 4H4v1.414z"
          data-marker="Close.path"
        />
      </svg>
    </div>
    <div
      class="c3"
      data-marker="Popover.Header"
    />
    <div
      class="c4"
      data-marker="Popover.Content"
    />
  </div>
</span>
`;

exports[`Popover w/ design 1`] = `
.c0 {
  outline: none;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  opacity: 0;
  width: 250px;
  min-width: 250px;
  box-shadow: 0 8px 32px 0 rgba(0,0,0,0.08);
  border-radius: 8px;
  padding-top: 12px;
  padding-right: 28px;
  padding-bottom: 12px;
  padding-left: 12px;
  z-index: 1000;
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
  background-color: #2B2D33;
  color: #FFFFFF;
  top: 22px;
  left: 0px;
  right: px;
  bottom: px;
  display: none;
}

.c5 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
  font-weight: 700;
  margin-bottom: 4px;
}

.c6 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
}

.c4 {
  position: absolute;
  top: 12px;
  right: 12px;
  cursor: pointer;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  outline: 0;
  position: relative;
  border-radius: 8px;
  border-color: transparent;
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.3s ease,color 0.3s ease,border-color 0.3s ease;
  transition: background-color 0.3s ease,color 0.3s ease,border-color 0.3s ease;
  font-size: 14px;
  line-height: 22px;
  padding: 0 20px;
  height: 36px;
  min-width: 36px;
  color: #FFFFFF;
  background-color: #2B2D33;
}

.c1:active {
  top: 1px;
}

.c1:hover {
  background-color: #000000;
  color: #FFFFFF;
}

.c2 {
  display: inline-block;
  vertical-align: top;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2:not(:last-of-type) {
  margin-right: 8px;
}

.c2 > svg {
  display: block;
}

<span
  class="c0"
  data-marker="Popover_0"
  tabindex="0"
>
  <button
    class="c1"
    data-marker="Button_0"
    type="button"
  >
    <span
      class="c2"
      data-marker="ButtonComponent.content.ContentWrapper"
    >
      Click me
    </span>
  </button>
  <div
    class="c3"
    data-marker="Popover.Message"
    width="250"
  >
    <div
      class="c4"
      data-marker="Popover.CloseIcon"
    >
      <svg
        data-marker="Popover.Close"
        fill="currentColor"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs
          data-marker="Close.defs"
        />
        <path
          d="M10.586 12L4 18.586V20h1.414L12 13.414 18.586 20H20v-1.414L13.414 12 20 5.414V4h-1.414L12 10.586 5.414 4H4v1.414z"
          data-marker="Close.path"
        />
      </svg>
    </div>
    <div
      class="c5"
      data-marker="Popover.Header"
    >
      This is header
    </div>
    <div
      class="c6"
      data-marker="Popover.Content"
    >
      Ut ut dui non ipsum pharetra aliquam id in libero. Aenean placerat leo nec ex pharetra finibus. Nunc id turpis eu risus rhoncus cursus quis efficitur sapien. Donec vitae arcu at lectus placerat fringilla.
    </div>
  </div>
</span>
`;

exports[`Popover w/ design 2`] = `
.c0 {
  outline: none;
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c3 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  opacity: 0;
  width: 250px;
  min-width: 250px;
  box-shadow: 0 8px 32px 0 rgba(0,0,0,0.08);
  border-radius: 8px;
  padding-top: 12px;
  padding-right: 28px;
  padding-bottom: 12px;
  padding-left: 12px;
  z-index: 1000;
  -webkit-transition: opacity .2s;
  transition: opacity .2s;
  background-color: #FFFFFF;
  color: #2B2D33;
  top: 22px;
  left: 0px;
  right: px;
  bottom: px;
  display: none;
}

.c5 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
  font-weight: 700;
  margin-bottom: 4px;
}

.c6 {
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-stretch: normal;
  font-style: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  margin: 0;
  font-size: 12px;
  line-height: 18px;
}

.c4 {
  position: absolute;
  top: 12px;
  right: 12px;
  cursor: pointer;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  cursor: pointer;
  outline: 0;
  position: relative;
  border-radius: 8px;
  border-color: transparent;
  font-family: 'ALS Hauss',Helvetica,'Helvetica New',Arial,sans-serif;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: background-color 0.3s ease,color 0.3s ease,border-color 0.3s ease;
  transition: background-color 0.3s ease,color 0.3s ease,border-color 0.3s ease;
  font-size: 14px;
  line-height: 22px;
  padding: 0 20px;
  height: 36px;
  min-width: 36px;
  color: #FFFFFF;
  background-color: #2B2D33;
}

.c1:active {
  top: 1px;
}

.c1:hover {
  background-color: #000000;
  color: #FFFFFF;
}

.c2 {
  display: inline-block;
  vertical-align: top;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c2:not(:last-of-type) {
  margin-right: 8px;
}

.c2 > svg {
  display: block;
}

<span
  class="c0"
  data-marker="Popover_0"
  tabindex="0"
>
  <button
    class="c1"
    data-marker="Button_0"
    type="button"
  >
    <span
      class="c2"
      data-marker="ButtonComponent.content.ContentWrapper"
    >
      Click me
    </span>
  </button>
  <div
    class="c3"
    data-marker="Popover.Message"
    width="250"
  >
    <div
      class="c4"
      data-marker="Popover.CloseIcon"
    >
      <svg
        data-marker="Popover.Close"
        fill="currentColor"
        height="16"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs
          data-marker="Close.defs"
        />
        <path
          d="M10.586 12L4 18.586V20h1.414L12 13.414 18.586 20H20v-1.414L13.414 12 20 5.414V4h-1.414L12 10.586 5.414 4H4v1.414z"
          data-marker="Close.path"
        />
      </svg>
    </div>
    <div
      class="c5"
      data-marker="Popover.Header"
    >
      This is header
    </div>
    <div
      class="c6"
      data-marker="Popover.Content"
    >
      Ut ut dui non ipsum pharetra aliquam id in libero. Aenean placerat leo nec ex pharetra finibus. Nunc id turpis eu risus rhoncus cursus quis efficitur sapien. Donec vitae arcu at lectus placerat fringilla.
    </div>
  </div>
</span>
`;
